README.md
# Multipass - PHP SDK

[![Build Status](https://travis-ci.org/SQweb-team/SQweb-SDK-PHP.svg?branch=master)](https://travis-ci.org/SQweb-team/SQweb-SDK-PHP)
[![Latest Stable Version](https://poser.pugx.org/sqweb/sdk_php/v/stable)](https://packagist.org/packages/sqweb/sdk_php)
[![Dependency Status](https://www.versioneye.com/user/projects/5650b42bff016c002c00056f/badge.svg)](https://www.versioneye.com/user/projects/5650b42bff016c002c00056f)

## Requirements

**This SDK has been tested with PHP 5.5 and greater.**

We are unable to provide official support for earlier versions. For more information about end of life PHP branches, [see this page](http://php.net/supported-versions.php).

## Install

**This package is intended for custom PHP websites and advanced integrations.**

If you're using WordPress, we've made it easy for you. Download the SQweb plugin [directly from WordPress.org](https://wordpress.org/plugins/sqweb/), or [check out the source](https://github.com/SQweb-team/SQweb-WordPress-Plugin).

### Recommended : Composer + dotenv

1. In your project root, execute `composer require mltpss/php-sdk`.

2. Create a `.env` file at the root of your project, or edit it if you already have one.

3. In `.env`, paste the following configuration:

    ```php
    SQW_ID_SITE=YOUR_WEBSITE_ID
    SQW_SITENAME=YOUR_WEBSITE_NAME
    SQW_LANG=en_US
    ```

    **Change `SQW_ID_SITE` with your website ID and `SQW_SITENAME` with the name to show on the large Multipass button**.

For additional settings, see "[Options](#options)" below.

### Composer + Manual Configuration

1. In your project root, execute `composer require mltpss/php-sdk`.
2. Create a new SQweb object passing in its configuration through the constructor. **`SQW_ID_SITE` and `SQW_SITENAME` must be specified**.

    ```php
    require_once('vendor/autoload.php');
    $sqweb = new \SQweb\SQweb(['SQW_ID_SITE' => 1234, 'SQW_SITENAME' => 'SQweb']);
    ```

For additional settings, see "[Options](#options)" below.

### Manually

1. Download [the latest release](https://github.com/SQweb-team/SQweb-SDK-PHP/releases) of the SDK and unzip it in a folder named `sqweb` in your project root.
2. Create a file named `sqweb.config` at the root of your project. This file should be one level up from the sqweb folder you just created, i.e. :

    ```text
    |–- sqweb/
    |   |-- src/
    |   |   |-- init.php
    |   |   |-- SQweb.php
    |-- sqweb.config
    ```

3. In `sqweb.config`, paste the following configuration:

    ```text
    SQW_ID_SITE=YOUR_WEBSITE_ID
    SQW_SITENAME=YOUR_WEBSITE_NAME
    SQW_LANG=en_US
    ```

    **Change `SQW_ID_SITE` with your website ID and `SQW_SITENAME` with the name to show on the large Multipass button**.

For additional settings, see "[Options](#options)" below.

## Usage

### 1. Initializing the SDK

First, you have to initialise SQweb variable on the pages where you wish to use it.

If you used composer:

```php
$sqweb = new SQweb\SQweb;
```

If you installed manually:

```php
include_once "whereYouInstalled/src/init.php";
```

### 2. Tagging your pages

Then, you need to add our JavaScript tag to your pages. This function will do it for you, and should be inserted before the closing `</body>` tag in your HTML.

```php
$sqweb->script();
```

Make sure it is present on all your pages. Most likely you'll just have to add it to your template.

**If you previously had a SQweb JavaScript tag, make sure to remove it to avoid any conflicts.**

### 3.a Checking the credits of your subscribers

Check if the user has credits, so that you can disable ads and/or unlock premium content.

```php
if ($sqweb->checkCredits() > 0) {
    // CONTENT
} else {
    // ADS
}
```

### 3.b Retrieving the user's email

If he wants it, the user can share his email with you, to retrieve it, check if the cookie named `mltpss_e` exists. Inside will be a base64 encoded string you need to retrieve and decode on your side.

for instance:
```php
if ($_COOKIE['mltpss_e']) {
    $email = base64_decode($_COOKIE['mltpss_e']);
}
```
This piece of code will store the email in the $email variable if the cookie is set.

### 4.a Showing the Multipass button

Use this code to display the Multipass button on your pages:

```php
$sqweb->button();
```

We have 3 additionals sizes for the button, to use it, pass a string to the function:

```php
$sqweb->button('tiny');
$sqweb->button('slim');
$sqweb->button('large');
```

We also have another type of button, the free button, which offer to share the user's email with you.
To display it, use this function:
```php
$sqweb->button('free');
```

![Example Buttons](https://cdn.multipass.net/github/buttons@2x.png "Example Buttons")

### 4.b Customizing the Multipass button

If you want to customize our different type of button, put the following in your config file (`.env` if you used composer and dotenv or `sqweb.config` if you did not) of in the sqweb config file.
For instance:
``SQW_LOGIN=Hello world``
Will show  `Hello world` instead of `Premium with Multipass` on the regular button for logged out visitors.


|Button Model|Logged in|Logged out
|---|:---:|:---:|
|Tiny|`SQW_CONNECTED_TINY`|`SQW_LOGIN_TINY`|
|Regular|`SQW_CONNECTED`|`SQW_LOGIN`|
|Large - before black dot| N / A | `SQW_BTN_UNLIMITED`|
|Large - after black dot| N / A | `SQW_BTN_NOADS`|
|Large connected| `SQW_CONNECTED_S` | N / A|

### 5. More functions

#### Display a support div for your users
```php
/**
 * Display a "Support us" message.
 */

function supportBlock() {   }
``

For instance:

```php
$sqweb = new SQweb;

$sqweb->supportBlock();
```

Will display the block.

#### Display a locking div for your users
```php
/**
 * Display a locking block.
 */

function lockingBlock() {   }
``

For instance:

```php
$sqweb = new SQweb;

$sqweb->lockingBlock();
```

We recommend that you use it with our other blocking function :

```php
if ($sqweb->waitToDisplay('15/09/16', 'd/m/y', 2)) {
    // The content here will appear 2 days after the publication date for non paying users.
} else {
    // Here you can put content that free users will see until the content above is available for all.
    $sqweb->lockingBlock();
}
```

#### Display only a part of your content to non premium users

```php
/**
 * Put opacity to your text
 * Returns text  with opacity style.
 * @param $text  Text you want to limit.
 * @param int $percent Percent of your text you want to show.
 * @return string
 */

function transparent($text, $percent = 100) { ... }
```

For instance:

```php
echo $sqweb->transparent('one two three four', 50);
```

Free users will see:

```text
one two
```

#### Display your content later for non paying users

```php
/*
 * @param string $date  When to publish the content on your site. It must be an ISO format(YYYY-MM-DD).
 * @param int $wait  Number of days you want to wait before showing this content to free users.
 */

function waitToDisplay($date, $wait = 0) { ... }
```

Example:

```php
if ($sqweb->waitToDisplay('15/09/16', 'd/m/y', 2)) {
    // The content here will appear 2 days after the publication date for non paying users.
} else {
    // Here you can put content that free users will see until the content above is available for all.
}
```

#### Limit the number of articles free users can read per day

For instance, if I want to display only 5 articles to free users:

```php
if ($sqweb->limitArticle(5) === true) {
    echo "This is my article";
} else {
    echo "Sorry, you reached the limit of pages you can see today, come back tomorrow or subscribe to Multipass to get unlimited articles !";
}
```

## Options

Unless otherwise noted, these options default to `false`. You can set them in your configuration.

|Option|Description
|---|---|
|`SQW_DEBUG`|Output various messages to the browser console while the plugin executes.|
|`SQW_DWIDE`|Set to `false` to only enable SQweb on the current domain. Defaults to `true`.|
|`SQW_LANG`|We support these locales: `en_US`, `en_GB` and `fr_FR`.|
|`SQW_AUTOLOGIN`|Enable or disable our autologin feature. Defaults to `true`.|

## Known Issues

If you change the value of `SQW_DWIDE` after initial deployment, your users will have to log in again since their auth cookie will no longer be valid.

## Troubleshooting

### I am not seeing the Multipass button

Please make sure you have properly [initialized the SDK](#1-initializing-the-sdk) and [tagged your pages](#2-tagging-your-pages). Mind the position of the include and tag.

### I can see the Multipass button, but I cannot login

The authentication requires cookies, which require [fully qualified domain names](https://en.wikipedia.org/wiki/Fully_qualified_domain_name). If you are testing on localhost or an IP, the authentication will succeed but won't be able to save a cookie in your browser.

## Contributing

We welcome contributions and improvements.

### Coding Style

All PHP code must conform to the [PSR2 Standard](http://www.php-fig.org/psr/psr-2/).

### Builds and Releases

See [RELEASE.md](RELEASE.md).

## Bugs and Security Vulnerabilities

If you encounter any bug or unexpected behavior, you can either report it on Github using the bug tracker, or via email at `hello@sqweb.com`. We will be in touch as soon as possible.

If you discover a security vulnerability within SQweb or this plugin, please send an e-mail to `hello@sqweb.com`. All security vulnerabilities will be promptly addressed.

## License

Copyright (C) 2015-2017 – SQweb

This program is free software ; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation ; either version 3 of the License, or any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY ; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details

You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
# Changelog

## v1.7.7 - 2018-07-18

* Improved the configuration setup

## v1.7.6 - 2018-07-17

* Added the free button

## v1.7.5 - 2018-04-16

* Fixed some typos.

## v1.7.4 - 2018-03-05

* Improved the way we log actions with the support

## v1.7.3 - 2018-03-01

* Fixed a wrong variable name.

## v1.7.2 - 2018-02-28

* Added a tunnel option.

## v1.7.1 - 2018-01-15

* Renamed the package from SQweb to Multipass

## v1.7.0 - 2018-01-09

* Now support more locales.
* Added autologin feature.
* Reworked script method.

## v1.6.2 - 2017-11-20

* Added a sqweb.config.example in assets directory.
* Now passing click event when opening iframe.

## v1.6.1 - 2017-11-07

* Implemented a 'support us' version of the login button.
* Made all logins button's text customizable.

## v1.6.0 - 2017-07-19

* BREAKING : Renamed misleading `sqweb_config.php` to `sqweb.config`.
* Added exemples pages which are also tests.

## v1.5.9 - 2017-07-11

* Fixed a typo in css.

## v1.5.8 - 2017-07-05

* Replaced the support block with a "Support us" message.

## v1.5.7 - 2017-06-08

* Added a new function to display a support block.
* Added a new function to display a locking block.

## v1.5.6 - 2017-05-26

* Added linebreaks cleanup to config file parser.

## v1.5.5 - 2017-05-24

* Fixed broken `explode()` in `sqweb_config.php` configuration option.

## v1.5.4 - 2017-05-23

* Fixed array syntax, for unsupported PHP versions.

## v1.5.3 - 2017-05-16

* Fixed typo in error message.
* Improved deployment automation.

## v1.5.2 - 2017-05-16

* Updated API endpoint to Multipass.net.

## v1.5.1 - 2017-05-16

* Introduced `SQW_ADBLOCK_MODAL` configuration variable.
* Updated CDN endpoint to Multipass.net.
* Fixed `_sqw.sitename` in config object (must be quoted).

## v1.5.0 - 2016-11-25

* Added `tiny` and `large` options for the Multipass button.
* Introduced `SQW_SITENAME` variable in our configuration file.
* Normalized variables names with our other SDKs.
* Leaner build system (replaced `gulp` with a shell script).

## v1.4.2 - 2016-11-01

* Added the option to display a smaller button.

## v1.4.1 - 2016-10-31

* Updated client script URL (now HTTPS only).

## v1.4 - 2016-10-11

* Fix bad call to sqwBalise

## v1.3 - 2016-09-23

* Added `transparent()` function to hide a part of your content.
* Added `limitArticle()` function to limit the number of articles read a day by free users.
* Added `waitToDisplay()` function to make the articles display later for free users.
* Readme improvements.

## v1.2 - 2016-06-07

* Improved `.env` handling.
* Added configuration via array.

## v1.1 - 2016-05-24

* Introduced autoloader support via Composer.
* Introduced `.env` configuration.
* Simplified manual installation.

## v1.0.4 - 2015-12-04

* Added versioning to Curl User Agent.

## v1.0.3 - 2015-12-03

* Improved documentation (Readme + comments).
* Added Gulp tasks for packaging.
* Updated Travis configuration.
* Clarified licensing.
* Added Changelog.

## v1.0.2 - 2015-11-27

* Updated the configuration to allow custom message definition.

## v1.0.1 - 2015-11-27

* Readme improvements.

## v1.0.0 - 2015-11-24

* Initial release.


tests/tests_buttons.php
<?php

include_once "../src/init.php";
?>

<!DOCTYPE html>
<html>
<head>
	<title>Testing multipass buttons</title>
</head>
<body>
	<h1>Here is a test page, it can serve as an example of how to display our buttons.</h1>
	<span style="color: #4491ff; float: left; width: 100%;">To check our filter tests, <a style="color: #4491ff" href="./tests_filters.php">click here</a></span>
	<span style="color: red">WARNING: All of these functions can only work properly if you used the function $sqweb->script() right before your &#60;/body&#61; tag </span>
	<h2>Here you should see the 4 buttons displayed from tiny to large with the corresponding function to display them.</h2>
	<div>
		<div style="display: inline-block;margin-right: 40px; margin-bottom: 20px; margin-top: 40px; width: 180px;"><pre>$sqweb->button('tiny');</pre></div>
		<div style="display: inline-block;"><?php $sqweb->button('tiny') ?></div>
	</div>
	<div>
		<div style="display: inline-block;margin-right: 40px; margin-bottom: 20px; width: 180px;"><pre>$sqweb->button('slim');</pre></div>
		<div style="display: inline-block;"><?php $sqweb->button('slim') ?></div>
	</div>
	<div>
		<div style="display: inline-block;margin-right: 40px; margin-bottom: 20px; width: 180px;"><pre>$sqweb->button('medium');</pre></div>
		<div style="display: inline-block;"><?php $sqweb->button('medium') ?></div>
	</div>
	<div>
		<div style="display: inline-block;margin-right: 40px; margin-bottom: 20px; width: 180px;"><pre>$sqweb->button('large');</pre></div>
		<div style="display: inline-block;"><?php $sqweb->button('large') ?></div>
	</div>
	<h2 style="margin-top: 50px;">Here you should see a "Support us" message with the corresponding function to display it.</h2>
	<div style="margin-top: 40px;"><pre>$sqweb->supportBlock();</pre></div>
	<div style="max-width: 1024px; padding: 60px; float:left;"><?php $sqweb->supportBlock(); ?></div>
	<?php $sqweb->script(); ?>
</body>
</html>

src/SQweb.php
<?php

/*
 * Multipass PHP SDK v1.7.7
 * @author Pierre Lavaux <pierre@multipass.net>
 * @author Mathieu Darrigade <mathieu@multipass.net>
 * @author Matthieu Borde <matthieu@sqweb.net>
 * @author Nicolas Verdonck <nicolas@sqweb.com>
 * @author Bastien Botella <bastien@sqweb.com>
 * @link https://github.com/SQweb-team/SQweb-SDK-PHP
 * @license http://opensource.org/licenses/GPL-3.0
 */

/**
 * Usage:
 * <code>
 * $sqweb = new \SQweb\SQweb();
 * $sqweb = new \SQWeb\SQweb(['SQW_ID_SITE'=>1234]);
 *
 * $sqweb->script(); // outputs <script...>
 * $sqweb->button(); // outputs <img src....>
 * </code>
 */
namespace SQweb;

class SQweb
{

    private $response;

    private $settings = array(
        'SQW_ID_SITE' => null,
        'SQW_SITENAME' => '',
        'SQW_DEBUG' => 'false',
        'SQW_BEACON' => 'false',
        'SQW_DWIDE' => 'false',
        'SQW_LANG' => 'en_US',
        'SQW_LOGIN' => '',
        'SQW_SUPPORT' => '',
        'SQW_CONNECTED' => '',
        'SQW_BTN_NOADS' => '',
        'SQW_LOGIN_TINY' => '',
        'SQW_CONNECTED_S' => '',
        'SQW_BTN_UNLIMITED' => '',
        'SQW_CONNECTED_TINY' => '',
        'SQW_CONNECTED_SUPPORT' => '',
        'SQW_AUTOLOGIN' => 'true',
        'SQW_TUNNEL' => 'support',
    );

    /**
     * @param array $config (optional) with keys matching the SQW_* class attributes.
     */
    public function __construct($config = array())
    {
        $this->loadConfig($config);
    }

    private function loadConfig($config)
    {
        if (!empty($config)) {
            if (empty($config['SQW_ID_SITE'])) {
                throw new InvalidArgumentException('SQW_ID_SITE must be defined.');
            }
            foreach (array_keys($this->settings) as $key) {
                if (array_key_exists($key, $config)) {
                    $this->settings[$key] = $config[$key];
                }
            }

            return true;
        }

        // Or use dotenv
        if (file_exists(__DIR__ . '/../../../../.env')) {
            $env = new \Dotenv\Dotenv(__DIR__ . '/../../../..');
            $env->load();
            $env->required('SQW_ID_SITE')->notEmpty();
            foreach (array_keys($this->settings) as $key) {
                if (getenv($key) !== false) {
                    $this->settings[$key] = getenv($key);
                }
            }

            return true;
        }

        // Or fallback to sqweb.config
        if (file_exists(__DIR__ . '/../../sqweb.config')) {
            $lines = file(__DIR__ . '/../../sqweb.config');
            foreach ($lines as $line) {
                $tmp = explode('=', $line);
                $key = $tmp[0];
                if (in_array($key, array_keys($this->settings))) {
                    $this->settings[$key] = rtrim($tmp[1]);
                }
            }

            return true;
        }
    }

    /**
     * Query the API for auth and credits status.
     * Returns the credits, or 0.
     * @return int
     */
    public function checkCredits()
    {
        if (empty($this->response)) {
            if (isset($_COOKIE['sqw_z']) && null !== $this->settings['SQW_ID_SITE']) {
                $curl = curl_init();
                curl_setopt_array($curl, array(
                    CURLOPT_URL => 'https://api.multipass.net/token/check',
                    CURLOPT_RETURNTRANSFER => true,
                    CURLOPT_CONNECTTIMEOUT_MS => 1000,
                    CURLOPT_TIMEOUT_MS => 1000,
                    CURLOPT_USERAGENT => 'SQweb/SDK_PHP 1.7.0',
                    CURLOPT_POSTFIELDS => array(
                        'token' => $_COOKIE['sqw_z'],
                        'site_id' => $this->settings['SQW_ID_SITE'],
                    ),
                ));
                $response = curl_exec($curl);
                curl_close($curl);

                $this->response = json_decode($response);
            }
        }
        if ($this->response !== null && $this->response->status === true && $this->response->credit > 0) {
            return $this->response->credit;
        }
        return 0;
    }

    /**
     * Output the JavaScript tag with its configuration object.
     */
    public function script($display = true)
    {
        $settings = json_encode(array(
            'wsid' => $this->settings['SQW_ID_SITE'],
            'sitename' => $this->settings['SQW_SITENAME'],
            'debug' => $this->settings['SQW_DEBUG'],
            'beacon' => $this->settings['SQW_BEACON'],
            'dwide' => $this->settings['SQW_DWIDE'],
            'locale' => $this->settings['SQW_LANG'],
            'autologin' => $this->settings['SQW_AUTOLOGIN'],
            'tunnel' => $this->settings['SQW_TUNNEL'],
            // User's custom strings for button customization
            'user_strings' => array(
                'login' => $this->settings['SQW_LOGIN'],
                'login_tiny' => $this->settings['SQW_LOGIN_TINY'],
                'connected' => $this->settings['SQW_CONNECTED'],
                'connected_tiny' => $this->settings['SQW_CONNECTED_TINY'],
                'connected_s' => $this->settings['SQW_CONNECTED_S'],
                'connected_support' => $this->settings['SQW_CONNECTED_SUPPORT'],
                'btn_unlimited' => $this->settings['SQW_BTN_UNLIMITED'],
                'btn_noads' => $this->settings['SQW_BTN_NOADS'],
                'support' => $this->settings['SQW_SUPPORT'],
            ),
        ));

        $output = '<script src="https://cdn.multipass.net/mltpss.min.js" type="text/javascript"></script>' . PHP_EOL;
        $output .= "<script>var mltpss = new Multipass.default($settings);</script>";

        if ($display == false) {
            return $output;
        }

        echo $output;
    }

    /*
     * Display a button for locked content
     */
    public function lockingBlock()
    {
        $html = $this->returnBlock('locking');

        echo $html;
    }

    /*
     * Display a supporting button
     */
    public function supportBlock()
    {
        switch ($this->settings['SQW_LANG']) {
            case 'fr':
            case 'fr_fr':
                $wording = array(
                    'title'         => 'L\'article est terminé ...',
                    'sentence_1'    => '... mais nous avons besoin que vous lisiez ceci: nous avons de plus en plus
                    de lecteurs chaque jour, mais de moins en moins de revenus publicitaires.',
                    'sentence_2'    => 'Nous souhaitons laisser notre contenu accessible à tous. Nous sommes
                    indépendants et notre travail de qualité prend beaucoup de temps, d\'argent et de dévotion',
                    'sentence_3'    => 'Vous pouvez nous soutenir avec Multipass qui permet de payer pour un bouquet de
                    sites, et ainsi financer le travail des créateurs et journalistes que vous aimez.',
                    'support'       => 'Soutenez nous avec'
                );
                break;

            default:
                $wording = array(
                    'title'         => 'Continue reading...',
                    'sentence_1'    => '... we need you to hear this: More people are reading our website than ever but
                    advertising revenues across the media are falling fast.',
                    'sentence_2'    => ' We want to keep our content as open as we can. We are independent,
                    and our quality work takes a lot of time, money and hard work to produce. ',
                    'sentence_3'    => 'You can support us with Multipass which enables you to pay for a bundle of
                    websites: you can finance the work of journalists and content creators you love.',
                    'support'       => 'Support us with'
                );
                break;
        }

        $html = '<div class="sqw-article-footer-container sqw-support-us-footer">
            <div class="sqw-article-footer-body">
            <div class="sqw-article-footer-body-title">' . $wording['title'] . '</div>
            <div class="sqw-article-footer-body-content1">' . $wording['sentence_1'] .'</div>
            <div class="sqw-article-footer-body-content2">' . $wording['sentence_2'] . '</div>
            <div class="sqw-article-footer-body-content3">' . $wording['sentence_3'] . '</div>
            </div>
            <div onclick="mltpss.modal_first(event)" class="sqw-article-footer-footer">
            <div class="sqw-article-footer-footer-text">' . $wording['support'] . '</div>
            <div class="sqw-article-footer-footer-logo-container"></div>
            </div>
            </div>';

        echo $html;
    }

    /*
     * Return the good button according to parent function.
     */
    public function returnBlock($type)
    {
        $wording = $this->selectText($type);

        return '
            <div class="footer__mp__normalize footer__mp__button_container sqw-paywall-button-container">
            <div class="footer__mp__button_header">
            <div class="footer__mp__button_header_title">' . $wording['warning'] . '</div>
            <div onclick="mltpss.modal_first(event)" class="footer__mp__button_signin">'
            . $wording['already_sub']
            . '<span class="footer__mp__button_login footer__mp__button_strong">'
            . $wording['login']
            . '</span></div>
            </div>
            <div onclick="mltpss.modal_first(event)" class="footer__mp__normalize footer__mp__button_cta">
            <a href="#" class="footer__mp__cta_fresh">' . $wording['unlock'] . '</a>
            </div>
            <div class="footer__mp__normalize footer__mp__button_footer">
            <p class="footer__mp__normalize footer__mp__button_p">'. $wording['desc'] . '</p>
            <a target="_blank" class="footer__mp__button_discover footer__mp__button_strong" href="'
            . $wording['href']
            . '"><span>></span> <span class="footer__mp__button_footer_txt">'
            . $wording['discover']
            . '</span></a>
            </div>
            </div>';
    }

    public function selectText($type)
    {
        $lang = $this->settings['SQW_LANG'];

        if ($type == 'support') {
            switch ($lang) {
                case 'fr':
                case 'fr_fr':
                    $wording = array(
                        'warning'       => 'Surfez sans publicité.',
                        'already_sub'   => 'Déjà abonné ? ',
                        'login'         => 'Connexion',
                        'unlock'        => 'Soutenez notre site grâce à ',
                        'desc'          => 'L\'abonnement multi-sites, sans engagement.',
                        'href'          => 'https://www.multipass.net/fr/sites-partenaires-premium-sans-pub-ni-limites',
                        'discover'      => 'Découvrir les partenaires'
                    );
                    break;

                default:
                    $href = 'https://www.multipass.net/en/premium-partners-website-without-ads-nor-restriction';
                    $wording = array(
                        'warning'       => 'Surf our website ad free',
                        'already_sub'   => 'Already a member? ',
                        'login'         => 'Sign in',
                        'unlock'        => 'Support our website, get your',
                        'desc'          => 'The multisite subscription, with no commitment.',
                        'href'          => $href,
                        'discover'      => 'Discover all the partners'
                    );
                    break;
            }
        } elseif ($type == 'locking') {
            switch ($lang) {
                case 'fr':
                case 'fr_fr':
                    $wording = array(
                        'warning'       => 'Cet article est reservé.',
                        'already_sub'   => 'Déjà abonné ? ',
                        'login'         => 'Connexion',
                        'unlock'        => 'Débloquez ce contenu avec',
                        'desc'          => 'L\'abonnement multi-sites, sans engagement.',
                        'href'          => 'https://www.multipass.net/fr/sites-partenaires-premium-sans-pub-ni-limites',
                        'discover'      => 'Découvrir les partenaires'
                    );
                    break;

                default:
                    $href = 'https://www.multipass.net/en premium-partners-website-without-ads-nor-restriction';
                    $wording = array(
                        'warning'       => 'The rest of this article is restricted.',
                        'already_sub'   => 'Already a member? ',
                        'login'         => 'Sign in',
                        'unlock'        => 'Unlock this content, get your ',
                        'desc'          => 'The multisite subscription, with no commitment.',
                        'href'          => $href,
                        'discover'      => 'Discover all the partners'
                    );
                    break;
            }
        }
        return $wording;
    }

    /**
     * Create the target button div.
     * @param null $size
     */
    public function button($size = null)
    {
        if ($size === 'tiny') {
            echo '<div class="sqweb-button multipass-tiny"></div>';
        } elseif ($size === 'slim') {
            echo '<div class="sqweb-button multipass-slim"></div>';
        } elseif ($size === 'large') {
            echo '<div class="sqweb-button multipass-large"></div>';
        } elseif ($size === 'support') {
            echo '<div class="sqweb-button-support"></div>';
        } elseif ($size === 'free') {
            echo '<div class="sqweb-button multipass-free"></div>';
        } else { // multipass-regular
            echo '<div class="sqweb-button"></div>';
        }
    }

    public function sqwBalise($balise, $match)
    {
        if (preg_match('/<(\w+)(?(?!.+\/>).*>|$)/', $match, $tmp)) {
            if (!isset($balise)) {
                $balise = array();
            }
            $balise[] = $tmp[1];
        }
        foreach ($balise as $key => $value) {
            if (preg_match('/<\/(.+)>/', $value, $tmp)) {
                unset($balise[ $key ]);
            }
        }
        return $balise;
    }

    /**
     * Put opacity to your text
     * Returns the text with opcaity style.
     * @param text, which is your text.
     * @param int percent which is the percent of your text you want to show.
     * @return string
     */
    public function transparent($text, $percent = 100)
    {
        if (self::checkCredits() === 1 || $percent == 100 || empty($text)) {
            return $text;
        }
        if ($percent === 0) {
            return '';
        }
        $arr_txt = preg_split('/(<.+?><\/.+?>)|(<.+?>)|( )/', $text, 0, PREG_SPLIT_NO_EMPTY | PREG_SPLIT_DELIM_CAPTURE);
        foreach (array_keys($arr_txt, ' ', true) as $key) {
            unset($arr_txt[ $key ]);
        }
        $arr_txt = array_values($arr_txt);
        $words = count($arr_txt);
        $nbr = ceil($words / 100 * $percent);
        $lambda = 1 / $nbr;
        $alpha = 1;
        $begin = 0;
        $balise = array();
        while ($begin < $nbr) {
            if (isset($arr_txt[$begin + 1])) {
                if (preg_match('/<.+?>/', $arr_txt[ $begin ], $match)) {
                    $balise = self::sqwBalise($balise, $match[0]);
                    $final[] = $arr_txt[ $begin ];
                    $nbr++;
                } else {
                    $tmp = number_format($alpha, 5, '.', '');
                    $final[] = '<span style="opacity: ' . $tmp . '">' . $arr_txt[ $begin ] . '</span>';
                    $alpha -= $lambda;
                }
            }
            $begin++;
        }
        foreach ($balise as $value) {
            $final[] = '</' . $value . '>';
        }
        $final = implode(' ', $final);
        return $final;
    }


    /**
     * Limit the number of articles free users can read per day.
     * @param $limitation int The number of articles a free user can see.
     * @return bool
     */
    public function limitArticle($limitation = 0)
    {
        if (self::checkCredits() == 0 && $limitation != 0) {
            if (!isset($_COOKIE['sqwBlob']) || (isset($_COOKIE['sqwBlob']) && $_COOKIE['sqwBlob'] != -7610679)) {
                $ip2 = ip2long($_SERVER['REMOTE_ADDR']);
                if (!isset($_COOKIE['sqwBlob'])) {
                    $sqwBlob = 1;
                } else {
                    $sqwBlob = ($_COOKIE['sqwBlob'] / 2) - $ip2 - 2 + 1;
                }
                if ($limitation > 0 && $sqwBlob <= $limitation) {
                    $tmp = ($sqwBlob + $ip2 + 2) * 2;
                    setcookie('sqwBlob', $tmp, time()+60*60*24);
                    return true;
                } else {
                    setcookie('sqwBlob', -7610679, time()+60*60*24);
                }
            }
            return false;
        }
        return true;
    }

    /**
     * Display your premium content at a later date to non-paying users.
     * @param  string  $date  Date of publishing the content on your site. It must be an ISO format(YYYY-MM-DD).
     * @param  integer $wait  Days to wait before showing this content to free users.
     * @return bool
     */
    public function waitToDisplay($date, $wait = 0)
    {
        if ($wait === 0 || self::checkCredits() === 1) {
            return true;
        }
        $format = 'Y-m-d';
        $datetime1 = new \Datetime;
        $datetime2 = $datetime1->createFromFormat($format, $date);
        $gap = (int)$datetime1->diff($datetime2)->format('%R%a');
        return ($gap + $wait) > 0 ? false : true;
    }
}
